Massive Dungeon Content Update

Zanaras opened the pull request:
20+ new monster types
5 new dungeon types
10 new classes for later implementation
Reduced minimum party size, with a chance that parties smaller than 3 will have a much harder start
Increase maximum party size, with additions to the code so they'll start off significantly harder
Upped the percentage of players the game will wait for to choose an action.

Honestly, a 30 man party on floor 1 should probably be WAY harder than it is now, based on my calculations, but I figured I'd let people give their reactions and work from that.

Anyways, I wanted to submit this so I could show some significant progress being made now that the code is public. That and I don't want to keep working on this on my own so long that by the time I submit it everything else has to be altered.

Commit:
Suicide runs, Raids, so much fun!

Worked through about line 260. Most of this makes sense so far. Well, up to line 260 anyways. In theory, I know what this does, sorta, but I'm not certain.

Commit:
Update DungeonMaster.php

Commit:
Update DungeonMaster.php

Commit:
Update DungeonCreator.php

Commit:
Update DungeonMaster.php

Commit:
Update dungeons.en.yml

Added new dungeon type descriptors and new monster sizes.

Commit:
Create TODO.md

Just a notes file for what I'm working on with this. --Andrew

Commit:
Update TODO.md

Commit:
Update TODO.md

Commit:
Update DungeonCreator.php

Added functionality for new dungeon types, began work on BossMonster selector, added indiv class usage to monster spawner,

Commit:
Update DungeonCreator.php

Commit:
Update MonsterClasses.md

explaining new classes

Commit:
Update MonsterClasses.md

Commit:
Update MonsterClasses.md

Commit:
Update LoadDungeonMonsters.php

23 new monsters, added in spawning for new dungeon types, added a slew of new classes in lieu of future work.

Commit:
Update dungeons.en.yml

Updating for new monsters and such.

Commit:
Update TODO.md

Just some plans.

Commit:
Create changes.md

Commit:
Update LoadDungeonMonsters.php

nerfed slimes, increased spawn rate!

Commit:
Update dungeons.en.yml

fixing typos

Commit:
Update dungeons.en.yml

Commit:
Update dungeons.en.yml

Missed a few entries. Should be the last time I have to edit this before the pull request.

Commit:
Update dungeons.en.yml

Commit:
Update dungeons.en.yml

WeaverofBrokenThreads commented on src/BM2/DungeonBundle/Service/DungeonMaster.php:
```diff
@@ -155,8 +155,8 @@ public function joinDungeon(Dungeoneer $dungeoneer, Dungeon $dungeon) {
 		list($party, $missing, $wait) = $this->calculateTurnTime($dungeon);
-		if ($party<3) return; // don't start until we have at least 3 people
-		if ($missing > $party/2) return; // don't start until at least half the members have selected an action
+		if ($party<1) return; // party must consist of at least one dungeoneer to begin
```
You could remove the line entirely.

Zanaras replied:
The game, if I understand it correctly, has a 0 member party exist in/at all dungeons, and I'd like to leave it so it can be used later for other things, like different dungeons having different minimum party sizes, without me having to sit down and figure out how to code it again.

Commit:
Update LoadDungeonMonsters.php

Adding the Mortal Dungeoneer ('npcdungeoneer') and Monster Hunter ('monhunter') that were contributed by cred from the forums.

Commit:
Update dungeons.en.yml

Adding the Mortal Dungeoneer ('npcdungeoneer') and Monster Hunter ('monhunter') that were contributed by cred from the forums.

Commit:
Update TODO.md

Just some notes for future work.

Zanaras commented:
Going to add a note to ensure that the dungeon closing logic is altered in my next pull request to accommodate larger party sizes. I don't foresee anyone getting enough players together right now to trigger a super early close before the next pull request, though. Parties of up to 10 dungeoneers won't be affected with increased modifiers in this pull request, and it'll be a while before I do a pull requests that adds minimum party sizes to certain dungeons.

Commit:
Update TODO.md

Just another reminder.

Commit:
Update DungeonCreator.php

Adding suggestions: flooded village, mausoleum, and ship graveyard; to the dungeon spawning list, as suggested by foxglove from the forums.

Commit:
Update LoadDungeonMonsters.php

Adding monster area logic for flooded village, mausoleum, and ship graveyard. Nerfed the slime slightly. Will add a card here soon to more easily slay them (especially since I plan on adding more of them, because I'm evil).

Commit:
Update dungeons.en.yml

Adding translations for the new dungeons.

Commit:
Update DungeonCreator.php

Removing boss monster code. I'll add it back in later once I actually start on that. Probably going to be with the dungeon type code.

Commit:
Update DungeonMaster.php

Overhauled the dungeon closing logic. 3 is minimum, 8 is average, 15 is max.

Commit:
Update map.js

Adding dungeon images. Not the greatest, but they work for now. Better images will be wanted for later though.

Commit:
Update dungeons.en.yml

Adding Slay Slime card.

Commit:
Update DungeonMaster.php

Adding the Slay Slime card.

Commit:
Update DungeonMaster.php

Making the slime card less useful on non-slimes.

Commit:
Update LoadDungeonCards.php

Adding Slay Slime card.

Commit:
Update LoadDungeonMonsters.php

Nerfed slimes slightly, to bring them in line with the slay slime card.

Commit:
Update changes.md

Commit:
Update dungeons.en.yml

tvogt commented on src/BM2/DungeonBundle/DataFixtures/ORM/LoadDungeonCards.php:
```diff
@@ -45,10 +47,10 @@ class LoadDungeonCards extends AbstractFixture implements OrderedFixtureInterfac
 		'fight.kill'		=> array('rarity' =>  100, 'monsterclass'=> '', 'target'=>array('monster'=>true, 'treasure'=>false, 'dungeoneer'=>false)),
 		'fight.sure'		=> array('rarity' =>  100, 'monsterclass'=> '', 'target'=>array('monster'=>true, 'treasure'=>false, 'dungeoneer'=>false)),
-		'fight.tripple'	=> array('rarity' =>  100, 'monsterclass'=> '', 'target'=>array('monster'=>true, 'treasure'=>false, 'dungeoneer'=>false)),
+		'fight.tripple'		=> array('rarity' =>  100, 'monsterclass'=> '', 'target'=>array('monster'=>true, 'treasure'=>false, 'dungeoneer'=>false)),
```
please keep indent on tabs. Otherwise you will have these changes in the diff which actually are no changes. It just makes it harder to understand what was actually changed.

Zanaras replied:
No changes on indenting were made in the file settings. All I did was was line them all up for easy comparison.

tvogt commented on src/BM2/DungeonBundle/DataFixtures/ORM/LoadDungeonMonsters.php:
```diff
@@ -13,30 +13,57 @@ class LoadDungeonMonsters extends AbstractFixture implements OrderedFixtureInter
+		'troglodyte'	=> array('power' => 80, 'attacks' =>  2, 'defense' => 50, 'wounds' => 3, 'mindepth' => 4, 'class' => array('monster', 'melee'), 'areas' => array('cave', 'wild', 'lab', 'flooded', 'shipgrave')),
+		'wyvern'	=> array('power' =>150, 'attacks' =>  2, 'defense' => 80, 'wounds' => 2, 'mindepth' => 5, 'class' => array('monster', 'fly'), 'areas' => array('wild', 'lab', 'flooded', 'shipgrave')),
+	
+		// First One Fortress
```
I like everything except this. First Ones should **always** be player characters.

tvogt commented on src/BM2/DungeonBundle/Resources/translations/dungeons.en.yml:
```diff
@@ -10,7 +10,9 @@ enter:
-   until you beat it or it beats you. Teamwork is key.
+   until you beat it or it beats you. Teamwork is key. <br />
+   Though, if you feel particularly daring, you can brave the worst
+   this dungeon has to offer on your own.
```
absolute no. I do not want dungeons to be solo-able. Never, ever. We can discuss lowering the minimum characters required to enter a dungeon, but I will not accept a pull request where it is one.

tvogt commented on src/BM2/DungeonBundle/Service/DungeonCreator.php:
```diff
@@ -76,8 +76,38 @@ public function createRandomLevel(Dungeon $dungeon, $depth) {
+		}
+
+		if ($members>30) {
+			// treat dungeon as another level deeper for parties of 30 --Andrew
```
the code above can be simplified as
`$depth += floor($members/5);`

tvogt commented on src/BM2/DungeonBundle/Service/DungeonMaster.php:
```diff
@@ -297,7 +297,7 @@ public function calculateTurnTime(Dungeon $dungeon) {
 		}
 		return array($party, $missing, $wait);
-	}
+	} // QUERY: what is the default wait time in real time? How do these numbers correlate to the real-world timing of things? --Andrew
```
the numbers above are real time **hours**.

tvogt commented:
Looks nice. As written in the in-line comments, please remove the First Ones fort, First Ones should always be player characters. Then I'll accept the PR.

Commit:
Update LoadDungeonMonsters.php

Removing first one as NPCs.

Commit:
Update DungeonCreator.php

Commit:
Update dungeons.en.yml

Commit:
Update map.js

Commit:
Update DungeonMaster.php

Commit:
Update dungeons.en.yml

Zanaras commented:
No more first ones as NPCs, reworked them and their fort to be mortals. Also nerfed the lords so the only things as strong as or stronger than first ones is actual monsters (though as is with combat, mortals can get really close).

The depth code I'm leaving as is, because I'm going to have to tweak it later. I'm pretty sure it's easier than it needs to be, but I've not got around to actually setting up a test database.

Minimum has been set to 2 for party size. There's a topic on the forum where I discuss some of my logic on it.

tvogt commented on src/BM2/DungeonBundle/Service/DungeonMaster.php:
```diff
@@ -155,8 +155,8 @@ public function joinDungeon(Dungeoneer $dungeoneer, Dungeon $dungeon) {
 		list($party, $missing, $wait) = $this->calculateTurnTime($dungeon);
-		if ($party<3) return; // don't start until we have at least 3 people
-		if ($missing > $party/2) return; // don't start until at least half the members have selected an action
+		if ($party<1) return; // party must consist of at least one dungeoneer to begin. Down from 3. --Andrew
```
This still needs to be changed to 2.

Commit:
Update DungeonMaster.php

Thinking about it, I've no idea why this wasn't done earlier.

Zanaras commented:
I did you one better. Instead of "$party<1" it's now more in line with the other place the minimum party size is referenced by being "$party < $this->min_party_size"

tvogt commented:
Done. Now I need to merge it into my local repository and then it can be deployed.

tvogt commented:
On local merging I got some errors about trailing whitespaces, probably in the .yml files.

tvogt commented:
There was a parse error in DungeonCreator.php line 79. The correct line is:
`if ($members<3 && rand(0,100) < rand(0,50)) {`
